### PR TITLE
[WIP] Implement build command.

### PR DESCRIPTION
**Pull request by ekse**
I started work on an implementation of the build command, based on the description in issue #8. I moved some of the code used by both replay and build in util.rs.

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Argh sorry will try to take a look at this tomorrow!

File: src/build.rs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                             &updated_tree,
+                             parents.as_slice());
+
+    match result {
+        Ok(oid) => println!("Commit: {:?}", oid),
+        Err(e) => error!("Failed to create commit: {}", e),
+    };
+}
+
+pub fn build(args: &Args) {
```
Comment (nikomatsakis) [inline src/build.rs:110]: Nit: can you move this function to the top of the file? I find it easier to read if things are kind of "top-down"

File: src/build.rs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            error!("failed to find repository containing `{}`: {}",
+                   cargo_toml_path.display(),
+                   e)
+        }
+    };
+
+    let repo_dir = cargo_toml_path.parent().unwrap();
+
+    // Check that there are no are untracked .rs files that might affect the build.
+    check_untracked_rs_files(repo);
```
Comment (nikomatsakis) [inline src/build.rs:128]: I think there is a similar function in `cargo replay` -- can we share this code?

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
I left a few nits but honestly they're pretty minor. I'm just going to merge this change. Sorry for the delay in review! This last few weeks have been crazy for me.

**Comment by ekse** [comment]
Thanks, I'll do the proposed changes.

File: src/build.rs (hunk 2 above)
Comment (ekse) [reply src/build.rs:128]: I copied the two functions below. We check for different conditions on status. Maybe we could pass a function and an error message to be printed?

``` rust
pub fn check_clean(repo: &Repository) {
    let statuses = match repo.statuses(None) {
        Ok(s) => s,
        Err(err) => error!("could not load git repository status: {}", err),
    };

    let mut errors = 0;
    let dirty_status = Status::all() - STATUS_IGNORED;
    for status in statuses.iter() {
        if status.status().intersects(dirty_status) {
            let stderr = io::stderr();
            let mut stderr = stderr.lock();
            if let Some(p) = status.path() {
                writeln!(stderr, "file `{}` is dirty", p).unwrap();
            }
            errors += 1;
        }
    }
    if errors > 0 {
        error!("cannot run with a dirty repository; clean it first");
    }
}

fn check_untracked_rs_files(repo: &Repository) {
    let statuses = match repo.statuses(None) {
        Ok(s) => s,
        Err(err) => error!("could not load git repository status: {}", err),
    };

    let mut errors = 0;
    for status in statuses.iter() {
        if status.status().intersects(STATUS_WT_NEW) {
            if let Some(p) = status.path() {
                if p.ends_with("rs") {
                    let stderr = io::stderr();
                    let mut stderr = stderr.lock();
                    writeln!(stderr, "file `{}` is untracked", p).unwrap();
                    errors += 1;
                }
            }
        }
    }
    if errors > 0 {
        error!("there are untracked .rs files in the repository");
    }
}
```

Comment (nikomatsakis) [reply src/build.rs:128]: Hmm yeah I guess they are slightly different. Let's leave it for now.

**Pull request merged**